src/proverapi.cpp

```diff
@@ -1,18 +1,50 @@
 #include "proverapi.hpp"
 #include "nlohmann/json.hpp"
 #include "logger.hpp"
+#include <chrono>
+#include <sstream>
 
 using namespace Pistache;
 using json = nlohmann::json;
 
 void ProverAPI::postInput(const Rest::Request& request, Http::ResponseWriter response) {
+    // Try to acquire semaphore - reject immediately if at capacity
+    if (!request_limit.try_acquire()) {
+        LOG_INFO("Server at capacity, rejecting request");
+        response.send(Http::Code::Service_Unavailable, 
+                     "Server at capacity, please retry later",
+                     MIME(Text, Plain));
+        return;
+    }
+    
+    // Start timing the request
+    auto request_start = std::chrono::steady_clock::now();
+    
+    // RAII guard ensures semaphore is released even if exception is thrown
+    SemaphoreGuard guard(request_limit);
+    
+    bool success = true;
     try {
+        // Generate proof
```

Comment:
let's minimize the locked scope, e.g.,

```suggestion
        // Generate proof
        json j = ..;
        {
            SemaphoreGuard guard(request_limit);
            j = prover.startProve(request.body());
        }
```